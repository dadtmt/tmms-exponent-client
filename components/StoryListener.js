import R from 'ramda'
import React, { Component, PropTypes } from 'react'
import { Text, View } from 'react-native'
import gql from 'graphql-tag'

import { pageEditorId } from '../config'
import Crossroads from './Crossroads'

const subscribeToCreateCrossroad = gql`
subscription SubscribeToCreateCrossroad($crossroadFilter:CrossroadSubscriptionFilter) {
  subscribeToCrossroad(mutations: [createCrossroad], filter:$crossroadFilter) {
    mutation
    edge {
      node {
        id
        isReady
        choices {
          edges {
            node {
              id
              text
              made
            }
          }
        }
        testDices{
          edges{
            node{
              id
              details
              nbDices
              nbSides
              made
              master
              modifier
              text
              result
            }
          }
        }
        text
      }
    }
  }
}
`

const subscribeToUpdateCrossroad = gql`
subscription SubscribeToUpdateCrossroad($currentCrossroad:CrossroadSubscriptionFilter) {
  subscribeToCrossroad(mutations: [updateCrossroad], filter:$currentCrossroad) {
    mutation
    edge {
      node{
        id
        isReady
        text
        choices{
          edges{
            node{
              id
              text
              made
            }
          }
        }
        testDices{
          edges{
            node{
              id
              details
              nbDices
              nbSides
              made
              master
              modifier
              text
              result
            }
          }
        }
      }
      }
    }
  }

`

const subscribeToCreateChoice = gql`
subscription SubscribeToCreateChoice($choiceFilter:ChoiceSubscriptionFilter) {
  subscribeToChoice(mutations: [createChoice], filter:$choiceFilter) {
    mutation
  	edge {
      node {
        id
        made
        text
      }
    }
  }
}
`

const subscribeToCreateTestDice = gql`
subscription SubscribeToCreateTestDice($testDiceFilter:TestDiceSubscriptionFilter) {
  subscribeToTestDice(mutations: [createTestDice], filter:$testDiceFilter) {
    mutation
  	edge {
      node{
        id
        details
        nbDices
        nbSides
        made
        master
        modifier
        text
        result
      }
    }
  }
}
`

export const getCurrentCrossroadFromData = R.pipe(
  R.pathOr([], ['getPageEditor', 'crossroads', 'edges']),
  R.head,
  R.propOr({ isReady: false }, 'node')
)

export const getCrossroadIdFromProps = R.pipe(
  R.pathOr([], ['data', 'getPageEditor', 'crossroads', 'edges']),
  R.head,
  R.path(['node', 'id'])
)

class StoryListener extends Component {

  componentWillReceiveProps(newProps) {
    const { data } = newProps
    const crossroadId = getCrossroadIdFromProps(newProps)
    const lastCrossroadId = getCrossroadIdFromProps(this.props)

    if(!this.newCrossroadSubscription) {
      this.newCrossroadSubscription = data.subscribeToMore({
        document: subscribeToCreateCrossroad,
        updateQuery: (prev, {subscriptionData}) => R.set(
          R.lensPath(['getPageEditor', 'crossroads', 'edges']),
          [subscriptionData.data.subscribeToCrossroad.edge]
        )(prev),
        variables: {
          crossroadFilter: {
            pageEditorId: {
              eq: pageEditorId
            }
          }
        }
      })
    }
    if (!R.isNil(crossroadId) && lastCrossroadId !== crossroadId) {
      this.newChoiceSuscription = data.subscribeToMore({
        document: subscribeToCreateChoice,
        updateQuery: (prev, {subscriptionData}) => R.over(
            R.lensPath(['getPageEditor', 'crossroads', 'edges']),
            R.over(
              R.lensIndex(0),
              R.over(
                R.lensPath(['node', 'choices', 'edges']),
                R.append(subscriptionData.data.subscribeToChoice.edge)
              )
            )
          )(prev),
        variables: {
          choiceFilter: {
            crossroadId: {
              eq: crossroadId
            }
          }
        }
      })
      this.newTestDiceSuscription = data.subscribeToMore({
        document: subscribeToCreateTestDice,
        updateQuery: (prev, {subscriptionData}) => R.over(
            R.lensPath(['getPageEditor', 'crossroads', 'edges']),
            R.over(
              R.lensIndex(0),
              R.over(
                R.lensPath(['node', 'testDices', 'edges']),
                R.append(subscriptionData.data.subscribeToTestDice.edge)
              )
            )
          )(prev),
        variables: {
          testDiceFilter: {
            crossroadId: {
              eq: crossroadId
            }
          }
        }
      })
      this.newUpdateCrossroadSuscription = data.subscribeToMore({
        document: subscribeToUpdateCrossroad,
        updateQuery: (prev, {subscriptionData}) => R.over(
            R.lensPath(['getPageEditor', 'crossroads', 'edges']),
            R.set(
              R.lensIndex(0),
              subscriptionData.data.subscribeToCrossroad.edge
            )
          )(prev),
        variables: {
          testDiceFilter: {
            crossroadId: {
              eq: crossroadId
            }
          }
        }
      })
    }
    if (!R.isNil(crossroadId) && lastCrossroadId !== crossroadId) {
      this.newTestDiceSuscription = data.subscribeToMore({
        document: subscribeToCreateTestDice,
        updateQuery: (prev, {subscriptionData}) => R.over(
            R.lensPath(['getPageEditor', 'crossroads', 'edges']),
            R.over(
              R.lensIndex(0),
              R.over(
                R.lensPath(['node', 'testDices', 'edges']),
                R.append(subscriptionData.data.subscribeToTestDice.edge)
              )
            )
          )(prev),
        variables: {
          testDiceFilter: {
            crossroadId: {
              eq: crossroadId
            }
          }
        }
      })
    }
  }

  render() {
    const { data, makeChoice, resolveTest } = this.props
    const { loading } = data
    const crossroads = R.pathOr(
      { edges: [] },
      ['getPageEditor', 'crossroads']
    )(data)
    const { isReady } = getCurrentCrossroadFromData(data)
    return (<View>
      {loading && <Text>Loading...</Text>}
      {isReady ? <Text>A toi de jouer!</Text> : <Text>Attends ton tour...</Text>}
      <Crossroads
        crossroads={crossroads}
        makeChoice={makeChoice}
        resolveTest={resolveTest}
      />
    </View>)
  }
}

StoryListener.propTypes = {
  data: PropTypes.object.isRequired,
  makeChoice: PropTypes.func.isRequired,
  resolveTest: PropTypes.func.isRequired
}

export default StoryListener
