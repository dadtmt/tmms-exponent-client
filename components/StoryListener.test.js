import React from 'react'
import renderer from 'react-test-renderer'

import StoryListener, {
  getCrossroadIdFromProps,
  getCurrentCrossroadFromData
} from './StoryListener'

describe('getCurrentCrossroadFromData', () => {
  it('return current crossroad', () => {
    const node = {
      id: 'SOME_ID',
      isReady: true
    }
    const data = {
      getPageEditor: {
        crossroads: {
          edges: [ { node } ]
        }
      }
    }
    expect(getCurrentCrossroadFromData(data)).toEqual(node)
  })
  it('return a default not ready crossroad', () => {
    const defaultCrossroad = {
      isReady: false
    }
    const data = {
      getPageEditor: {
        crossroads: {
          edges: []
        }
      }
    }
    expect(getCurrentCrossroadFromData(data)).toEqual(defaultCrossroad)
  })
})

describe('getCrossroadIdFromProps', () => {
  it('should return crossroadId', () => {
    const props = {
      data: {
        getPageEditor: {
          crossroads: {
            edges: [
              {
                node: {
                  id: 'SOME_ID'
                }
              }
            ]
          }
        }
      }
    }
    expect(getCrossroadIdFromProps(props)).toBe('SOME_ID')
  })
  it('should return falsy when edges empty', () => {
    const props = {
      data: {
        getPageEditor: {
          crossroads: {
            edges: []
          }
        }
      }
    }
    expect(getCrossroadIdFromProps(props)).toBeFalsy()
  })
  it('should return falsy when no getPageEditor', () => {
    const props = {
      data: {}
    }
    expect(getCrossroadIdFromProps(props)).toBeFalsy()
  })
})

describe('StoryListener', () => {
  it('renders loading', () => {
    const props = {
      data: {
        loading: true
      },
      makeChoice: jest.fn(),
      resolveTest: jest.fn()
    }
    expect(
      renderer.create(<StoryListener {...props} />).toJSON()
    ).toMatchSnapshot()
  })

  it('renders current page', () => {
    const text = {
      blocks: [
        {
          data: {},
          depth: 0,
          entityRanges: [],
          inlineStyleRanges: [],
          key: '4bkmd',
          text: 'some rich text',
          type: 'unstyled'
        }
      ],
      entityMap: {}
    }
    const props = {
      data: {
        getPageEditor: {
          crossroads: {
            edges:[
              {
                node: {
                  choices: { edges: [] },
                  id: 'SOME_ID',
                  testDices: { edges: [] },
                  text
                }
              }
            ]
          }
        },
        loading: false
      },
      makeChoice: jest.fn(),
      resolveTest: jest.fn()
    }
    expect(
      renderer.create(<StoryListener {...props} />).toJSON()
    ).toMatchSnapshot()
  })

})
