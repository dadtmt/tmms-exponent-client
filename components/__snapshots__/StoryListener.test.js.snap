exports[`StoryListener renders current page 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    Attends ton tour...
  </Text>
  <View>
    <View>
      <View
        style={
          Object {
            "marginBottom": 20,
          }
        }>
        <View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "fontSize": 15,
              }
            }>
            some rich text
          </Text>
        </View>
      </View>
      <View />
      <View />
    </View>
  </View>
</View>
`;

exports[`StoryListener renders loading 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    Loading...
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail">
    Attends ton tour...
  </Text>
  <View />
</View>
`;
